Engine support portforward, send server port, try near ports, predict ports

The-any-Key opened the pull request:
<H1>New engine features:</H1>
**Common stuff:**
Variable to set in the config so you can set the version of gamemaker you use.

**If connection fail client try:**
Server port nr
Master server port nr
Ports near the received external port nr
Save external ports when use master server to predict ports

**Server can now easy:**
Send its port number to the connecting clients. If the server use port forward this will come in handy.

Parakoopa commented:
This has been merged into master/punch-improvements, so I can do some testing. I haven't really worked with PRs too much but I know github doesn't like merging them into other branches than originally selected.

But I think when I merge punch-improvements back into master this PR should close. Let's try it out.

Parakoopa commented:
You accidentally partially merged your upnp code into your master. I will instead merge "Reset to default values …	2746c0a" into punch-improvements.

The-any-Key commented:
Yes I fixed that in "Removed upnp from master" commit.

Parakoopa commented:
Yeah I've seen that but you forgot some files (eg. execute_program_dir.gml). No problem though, I've just merged the point before you merged your other branch in. If you know how you should propably roll back your master to this commit as well (2746c0a)

In case you don't here's how you do it:
    git checkout master
    git reset 2746c0a --hard
    git push origin master --force

Please note that you should never do this when somebody else works on the same repository as you because that will cause real trouble on the other people's end. But since you work on your fork alone, this is fine.

Parakoopa commented:
You also merged your master back into your upnp branch, which you also shouldn't do because now I can't really merge it without also merging your changes in this pull request. This might lead to trouble when I try to merge everything back together later.

But don't worry, git workflow can be a little bit confusing sometimes, especially if you are new to it, just some tips for the future

The-any-Key commented:
Ahh. Removed the script from the project but it was still in the script folder. Fixed now. Very new to github.

Parakoopa commented:
Cool!
But please consider just reseting as I showed you above, because just removing everything again with commit will leave the repository kinda dirty and it will lead to some issues when you try to send me other pull requests later on. It's kinda hard to explain and I'm not really good at doing so, but let's just say git get's confused if we decide to handle it this way.


What git client are you using? I guess you are using some kind of gui and I could help you execute the commands above with that.

The-any-Key commented:
I think I got it now.

Parakoopa commented:
Ah cool! Awesome :)

Parakoopa commented on GMnetENGINE.gmx/scripts/htme_config.gml:
```diff
@@ -41,6 +49,11 @@ self.udphp_master_ip = "95.85.63.183";
+// Send port to client on online server
+// You can manually port forward the server port to enable clients to connect
+// Note that this will use data8 to provide the port so dont set it yourself
+self.udphp_provide_server_port=true;
```
I don't really get why the entire udphp_provide_server_port logic is needed.
GMnet PUNCH automatically falls back to connecting directly with the server what is actually the problem here I think is that the lobby doesn't do this yet.

So instead of having this to be an option, this should just be added to GMnet without an option to turn it on or off.

The-any-Key replied:
Well I added it if someone use all the data variables and cant let go of data8. But it could just be added and reduce the available spots.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientPunch.gml:
```diff
@@ -63,9 +63,59 @@ if (!client_connected) {
+        {
+            // Try connect with provided port and hope the server got portforward
+            // The port may be provided by the server to the master and to us if
+            // global.htme_object.udphp_provide_server_port=true in config
```
The GMnet variable can't be used this way here since this is part of GMnet PUNCH, which is supposed to work standalone without GMnet ENGINE.
However don't worry! I can simply add/move the udphp_provide_server_port logic to GMnet PUNCH. 
This mainly means adding the udphp_provide_server_port variable to udphp_config.gml and adding functions for GMnet ENGINE to sync the udphp_provide_server_port port variable into that.

BUT please see my comment at htme_config.gml because this whole variable might not even be needed anyway.

Parakoopa commented on GMnetENGINE.gmx/scripts/htme_clientStart.gml:
```diff
@@ -37,6 +37,31 @@ if( self.socketOrServer<0 ){
+        {
+            var entry = l[| 0];
+            var ip = entry[? "ip"];
+            var port_provided = entry[? "data8"];
```
Instead of using data8 (which is reserved for the user) we should add a completely new variable. 

This means GATE.PUNCH needs to be changed as well.
I was actually going to throw the current GATE.PUNCH out of the window and redo it in nodejs since nodejs works so much better for this kind of purpose than java. I will also add a new field for that then.

The-any-Key replied:
Yes that's why I used a option to include it. But if you think of rebuilding a new gate punch in node.js.

*add to personal to do list:
1# Learn what node.js is
2# Learn some node.js

Parakoopa commented on GMnetENGINE.gmx/scripts/htme_clientStart.gml:
```diff
@@ -37,6 +37,31 @@ if( self.socketOrServer<0 ){
     return false;
 }
 
+// Get server port from list
```
This does not seem right here.
Shouldn't this instead be added to obj_udphphtme_lobby and the manual?
Have a look at line 29 of obj_udphphtme_lobby's event "KEY PRESS 1". We can just give the port there from the lobby entry instead of giving the port 0 to this script.

The-any-Key replied:
Yes. I tried to not to change the demo objects only the scripts so thats why this seems not right. I dont know why I am afraid of changing the demo objects :-)

Parakoopa commented on GMnetENGINE.gmx/scripts/htme_serverStart.gml:
```diff
@@ -46,6 +54,8 @@ if (self.use_udphp) {
     }
     script_execute(asset_get_index("udphp_serverSetData"),1,self.gamename);
+    // Send server port to master server so the client can get it and connect to it directly
+    if global.htme_object.udphp_provide_server_port script_execute(asset_get_index("udphp_serverSetData"),8,string(self.port));
```
See the comment regarding data8 in htme_clientStart.gml

Parakoopa commented:
I've added some first comments on the overall procedure (please have a look :) ). Haven't really looked into the new feature-code for punch yet, but I'll do that now.

But what I've seen so far looks very nice :)

Parakoopa commented on GMnetENGINE.gmx/scripts/htme_clientStop.gml:
```diff
@@ -22,6 +22,6 @@ with (global.htme_object) {
 htme_shutdown();
 if (self.use_udphp) {
-    script_execute(asset_get_index("udphp_stopClient"));
+    script_execute(asset_get_index("udphp_stopClient"),self.udphp_client_id);
```
Thank's!! That might solve a mayor memory leak.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientNetworking.gml:
```diff
@@ -54,6 +54,45 @@ if (in_ip == global.udphp_master) {
+                }
+            }
+            // Save values for next connect
+            ini_open("udphp_predict.ini");
```
Nice idea to store this information for later.

The-any-Key replied:
When the server restart and we connect again we will get a new external port and predict the next one.

This should actually be done by connecting to another master server (2) and send the external port to the server and then connect to master server (1) (with lobby) and send the external ip from master server (1) to master server (2). Now master server (1) got 2 external ports to send to the connecting client and we can predict the new port much better. The clients will do the same and send 2 external ports to the server. 

But I dont know if people want to setup 2 master server to do this.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientNetworking.gml:
```diff
@@ -54,6 +54,45 @@ if (in_ip == global.udphp_master) {
+            ini_open("udphp_predict.ini");
+            ini_write_real("predict-" + string(server_global_ip),"value1",global.udphp_punch_stage_predict_value1);
+            ini_write_real("predict-" + string(server_global_ip),"value2",global.udphp_punch_stage_predict_value2);
+            ini_close();            
```
It would propably be better to store all of this in a new script instead to keep it all more readable.

Parakoopa replied:
Wait. I don't really get what this part of script does. Isn't the port prediction done in the punchstate script? What is this for?

The-any-Key replied:
This script updates the external ports and save them to the ini. 
On first connect
global.udphp_punch_stage_predict_value1=55574 (New external port)
global.udphp_punch_stage_predict_value2=0
On second connect (if server restarted and got new external port)
global.udphp_punch_stage_predict_value1=55574
global.udphp_punch_stage_predict_value2=55570 (New external port)
On third connect (if server restarted and got new external port)
55560 is the new external port and we want to replace the value that is nearest this. So we replace value2=55570 to value2=55560
so:
global.udphp_punch_stage_predict_value1=55574
global.udphp_punch_stage_predict_value2=55560 

The globals is then used in the punchstate script.

Parakoopa replied:
Ah I see.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientPunch.gml:
```diff
@@ -63,9 +63,59 @@ if (!client_connected) {
+            // The port may be provided by the server to the master and to us if
+            // global.htme_object.udphp_provide_server_port=true in config
+            // Check if port is provided
+            if global.htme_object.server_port>0
```
The GMnet variable can't be used this way here since this is part of GMnet PUNCH, which is supposed to work standalone without GMnet ENGINE.

UDPHP should use it's own client port which will be synced automatically be GMnet ENGINE in htme_clientStart, so you can just use this port here.

The-any-Key replied:
Ahh.

Parakoopa replied:
Hi, I'm cleaning this up right as I type:

You set the client port of the PUNCH client to the port of ENGINE client.
As I said this should already be the same! - In this case this part can be removed. Was this not the case during your testings?

The-any-Key replied:
I know there was something fishy about this. I think that when I tested direct connect I run into problems with the port. But I don't remember what it was. We will see that later on when testing it more.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientPunch.gml:
```diff
@@ -63,9 +63,59 @@ if (!client_connected) {
+            global.udphp_punch_stage_sub1="";
+            global.udphp_punch_stage_counter=0;            
+            // Only config this once
+            global.udphp_punch_stage="Try sequence external server port";            
```
Instead of writing strings to udphp_punch_stage enums should be used.

Parakoopa commented on GMnetENGINE.gmx/scripts/udphp_clientPunch.gml:
```diff
@@ -63,9 +63,59 @@ if (!client_connected) {
+        }        
+        else if (client_timeout = floor(global_timeout*0.45) and global.udphp_punch_stage!="Try master server port")
+        {    
+            // Try connect with master port and hope the server got the same port and use portforward
```
I don't really see how this method would be useful, why should the client have the same port as the master server?

The-any-Key replied:
Sorry. I was on a must-try-everything-to-make-it-work approach. I was in a desperate state and thought - "why not, this could maybe happen" :-)

Parakoopa commented:
Hi, just a quick info:
I'm starting to clear some stuff up in your changes soon, hopefully today. If you want to get these changes you can merge the branch punch-improvements back into your master branch. If you have any changes that you'd like to share with me, just tell me here and I will merge your master back into punch-improvements.
This way we can work together on this pull request.

Parakoopa commented:
I have pushed some changes. These might break your changes; since you implemented the new features could you have a look at my changes and report back if everything still works. I also did some basic testing and everything seems to be working...

The-any-Key commented:
Ok. Got the new punch-improvements branch.

Parakoopa commented:
I've looked again at the "send port via data8" part and I think I missed something at first glimpse, which will make this a little bit more complicated.

I removed your "htme_object.udphp_provide_server_port" flag because I didn't realise that the client would always replace the actual port reported by the master server lobby with the port the server things he's running on.
We should handle this differently. Instead both htme_startClient and udphp_clientStart should allow a second port to be passed to the function, which will then be used with all of the old and new holepunching methods if the port reported by the master server fails.

This will need some planning and rewriting to do that smoothly.

Parakoopa commented:
Most of your new holepunching doesn't support multiple clients running at the same time. My advice is that we either...
1) just remove all multi-client support and only use one global variable per information instead of maps
2) make each client an actual GameMaker instance of an object which stores the information.

The-any-Key commented:
I am a little lost.
This is the concept when I created the code:
When use "htme_object.udphp_provide_server_port" the created server provide the internal port the server socket listen to. When a client try to connect he will try default punch with the external port from the master server, then try the internal server port provided, then near external ports, then try predict the port with old saved external ports. The internal provided port is only if the server have setup port-forward on his router. 

I remembered why I added the master server port now :-). Because if the server for some reason dont provide the internal socket port. The internal master server port is the same as the default internal server port. And Why not check if the server use portforward. When a user ask for help you can just tell him "try open the 6510 on the router" and it will start to work.

What problems do you run into when running multiple clients?

Parakoopa commented:
For example your punchstate variable is a single global variable.
In theory PUNCH can run multiple clients at the same time. Check out the demo project of GMnet PUNCH in the marketplace (https://marketplace.yoyogames.com/assets/411/multiplayer-without-open-ports-gmnet-punch).

However when only using one variable for controling the punchstate, all will use the same variable which won't work correctly. Instead this whole map concept has to be used (like how ports and ip's and the buffers are all stored in a ds_map)
But as I said this is pretty rubbish anyway. Either we drop multi-client support or have each PUNCH client be an actual GameMaker object.


As for the port-thing:
"When a user ask for help you can just tell him "try open the 6510 on the router" and it will start to work." - The idea in this case would be for the people creating the actual games to use one port instead of having the user manually specify the port. This is the usual thing to do. In your game you wouldn't actually ask the player to specify a port - or at least only make it optional to use another one than the default. So this isn't needed since the specific game WILL have a default game that you as a game developer can just say: Well if it doesn't work open port XYZ.

"When use "htme_object.udphp_provide_server_port" the created server provide the internal port the server socket listen to. " - I know. I get the idea, it's just that we have to integrate it into the actual thing itself and not do a "hack" like right now where the clientStart script checks if the server comes from a lobby and if so just overwrites the port - clientStart and udphp_startClient should have an optional argument to specify this port. And when you program your lobby you can just pass that through - PUNCH should then do all hole punching operations on the primary port first, and then on the optional port which should be the internal socket port provided by the lobby

The-any-Key commented:
Ah. ok. I was just thinking of the gmnet engine when use in a gm game. When the my punch variables is only used by a single client who try to connect to a server. I didnt think of using them in the gmnet punch. Ok that explains a lot. I am not so familiar with the gmnet punch project on what it does.

Understand your thinking of the port thingi. When I added that I wanted the programmer to be able to easy setup many dedicated servers (with gmnet engine) with different ports on the same machine so people could connect to one. Or when many people on the same NAT want to play, but not together and use portforward. Or when people want a private game on a "secret" port (I use filter when use on the online lobby). 
But I understand that the gmnet punch is the problem here. What is the fundamental of the gmnet punch project? You use it when you already have a network engine in a game and want to add punch abilities?

Sorry about the GMAssetCompiler.exe. I changed to gm so I could escape the manual compiler stuff.

Parakoopa commented:
Yes, GMnet PUNCH is meant to be used in all kinds of engines, either GMnet
ENGINE or other custom multiplayer solutions. To make development easier 
it's in
the same repository here on github, since GMnet ENGINE can be used to test 
it.
But your changes would also go into the stand-alone product GMnet PUNCH, 
that's
why I'd like to seperate PUNCH and ENGINE code. I'll figure out something 
about
the whole multi-client situation and adapt your changes.

The-any-Key commented:
Ok. That explains the confusion and the problem. :-)

Parakoopa commented:
Hi again!
Your Pull Request does not pass all tests of the new PUNCH testing suite*
http://parakoopa.de:8080/job/GMnet_PunchTest/42/testReport/

The test that fails is the test that tests connection without master server and a server running. That means it just starts a client and tries to connect to a server which doesn't exist.

The reason it fails is because the connection doesn't fail in 5 seconds (set via global_timeout variable). It takes longer; about 10 seconds. 

I looked through the code and I'm not sure why that happens. Maybe the game is lagging while it's trying to connect?

\* The suite currently only tests restricted/full cone NATs (NATs that support hole punching on same Port)

The-any-Key commented:
Yes. If you try to connect to a master server that is offline game maker hangs for about 5 sec.
I guess it has something with the tcp. I got this problem when I started using your engine and didn't have a master server. Someone asked on the forum why he got a delay on start and I sad he could just set self.use_udphp = false; if he didn't have a master server.

I tried to find the post but the forum dont let me search my posts for "use_udphp". it's to common it says. Is it possible to disable "ignore common words" so I can search for a common word on a specific set of posts? I tried to search for a common word on one post and it just say it's to common :laughing: 

when tested on Android I did not have the delay. So it may be windows related.

Parakoopa commented:
Hi, please create a new pull request based on a forked version of my punch-improvements branch (see message I sent you).

Parakoopa commented:
This PR is continued in #4
